### PR TITLE
Parser Performance: Cache segment string repr to reduce function calls

### PR DESCRIPTION
**Pull request by WittierDinosaur**
### Brief summary of the change made
The current version of __repr__ in raw.py doesn't cache the string. This means that every comparison causes a rebuild, and also recalculates the pos marker. This PR caches it as a property. I tested the parser performance of this PR on a large file (~2300 lines) using CProfile:

Before:
```
424944039 function calls (413269403 primitive calls) in 221.592 seconds
```
After
```
213425726 function calls (201751080 primitive calls) in 110.296 seconds
```

### Are there any other side effects of this change that we should be aware of?
Not that I'm aware of

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -34,6 +34,7 @@ test-reports
 # Ignore root testing sql & python files
 /test*.sql
 /test*.py
+/test*.txt
```
Comment (alanmcruickshank) [inline .gitignore:37]: Why this?

File: src/sqlfluff/core/parser/segments/raw.py [hunk 2]
```diff
@@ -58,12 +58,13 @@ def __init__(
...
-
-    def __repr__(self):
-        return "<{}: ({}) {!r}>".format(
+        self.representation = "<{}: ({}) {!r}>".format(
             self.__class__.__name__, self.pos_marker, self.raw
         )
 
+    def __repr__(self):
+        return self.representation
+
```
Comment (alanmcruickshank) [inline src/sqlfluff/core/parser/segments/raw.py:67]: Looks great 👍 . Can you add a short comment here to note the intent of `self.representation` for future-us? Feels like something we'll totally forget.

**Review by alanmcruickshank — COMMENTED** [review]
This looks like such an excellent idea! Good thinking.

Couple of notes on the details, but otherwise I think good to go.

File: .gitignore (hunk 1 above)
Comment (WittierDinosaur) [reply .gitignore:37]: Unrelated change - but when I use cProfile, I tend to output to .txt, so adding this (currently unused) namespace to give me somewhere in root to dump my outputs

File: src/sqlfluff/core/parser/segments/raw.py (hunk 2 above)
Comment (WittierDinosaur) [reply src/sqlfluff/core/parser/segments/raw.py]: Done 👍

File: .gitignore (hunk 1 above)
Comment (alanmcruickshank) [reply .gitignore:37]: Ok - I don't have a good reason for us to use txt files for much, so makes sense.

**Review by alanmcruickshank — APPROVED** [review]

**Pull request merged**